build/bazel/remote/execution/v2/remote_execution.proto

```diff
@@ -1288,6 +1670,228 @@ message WaitExecutionRequest {
+message ObservedPathSet
+{
+  // An individual file or directory access observed during execution of
+  // an [Action][build.bazel.remote.execution.v2.Action].
+  message Entry
+  {
+    // Operation kind.
+    enum OpKind
+    {
+      // `Path` refers to a file that was read by the action. If the file is
+      // not present, this pathset does not match the filesystem. Otherwise
+      // the hash of the string constant "R" followed by the hash of the file
+      // must be added to the observed-input digest.
+      FILE_READ = 0;
+
+      // `Path` is a directory that was enumerated by the action using
+      // `DirectorySearchPattern`. If the directory is missing, this pathset
+      // does not match the filesystem. Otherwise the hash of the string
+      // constant "E" followed by the hash of the directory contents with the
+      // search pattern applied must be added to the observed-input digest.
+      // See comments on
+      // [ObservedPathSet][build.bazel.remote.execution.v2.ObservedPathSet]
+      // for how to generate a directory hash.
+      DIRECTORY_ENUMERATION = 1;
+
+      // `Path` is a directory that was enumerated by the action. If the
+      // directory is missing, this pathset does not match the filesystem.
+      // Otherwise the hash of the string constant "E" followed by a hash of
+      // the directory contents filtered to only the file names in
+      // `ObservedAccessFileNames` (compared to only the directory's filenames
+      // without extensions) must be added to the observed-input digest.
+      // See comments on
+      // [ObservedPathSet][build.bazel.remote.execution.v2.ObservedPathSet]
+      // for how to generate a directory hash.
+      DIRECTORY_SEARCH = 2;
+
+      // 'Path' is a file probed for existence, but not read, by the action
+      // and that was present during execution. If the file is not present or
+      // is instead a directory, this pathset does not match the filesystem.
+      // Otherwise the hash of the string constant "F" must be
+      // added to the observed-input digest.
+      EXISTING_FILE_PROBE = 3;
+
+      // `Path` is a directory probed for existence, but not enumerated, by
+      // the action and that was present during execution. If the directory is
+      // not present, or is instead a file, this pathset does not match the
+      // filesystem. Otherwise the hash of the string constant "D" must be
+      // added to the observed-input digest.
+      EXISTING_DIRECTORY_PROBE = 4;
+
+      // `Path` is a file or directory probed for existence by the action
+      // that was not present during execution. If path is present as a file
+      // or directory, this pathset does not match the filesystem. Otherwise
+      // the hash of the string constant "P" must be added to the
+      // observed-input digest.
+      ABSENT_FILE_OR_DIRECTORY_PROBE = 5;
+    }
```

Comment:
This approach to specifying file operations seems underspecified. There are a number of properties, like the contents of `FileProperties` and `is_executable`, which could be distinguished under this API but which aren't encoded here. On unix, probably the correct thing to do is encode that a `stat()` call was made and list all of the queried properties.

I'm unsure about what should be done about properties that could be in `FileProperty` and would fall outside that, which I imagine could include Linux extended ACLs. I guess this should also accept extension operations, then.

Reply:
Discussing with the sandboxing devs. My first response would be that file presence and hash are sufficient, but maybe we've missed a nuance here.